nav2_costmap_2d/src/costmap_subscriber.cpp

```diff
@@ -64,40 +105,118 @@ std::shared_ptr<Costmap2D> CostmapSubscriber::getCostmap()
+        for (unsigned int x = 0; x < costmap_update_msg_->width; x++) {
+          unsigned int index = index_base + x + costmap_update_msg_->x;
+          master_array[index] = interpretValue(costmap_update_msg_->data[di++]);
+        }
+      }
+    } else {
+      if (costmap_ == nullptr) {
+        costmap_ = std::make_shared<Costmap2D>(
+          costmap_msg_->info.width, costmap_msg_->info.height,
+          costmap_msg_->info.resolution, costmap_msg_->info.origin.position.x,
+          costmap_msg_->info.origin.position.y);
+      } else if (costmap_->getSizeInCellsX() != costmap_msg_->info.width ||  // NOLINT
+        costmap_->getSizeInCellsY() != costmap_msg_->info.height ||
+        costmap_->getResolution() != costmap_msg_->info.resolution ||
+        costmap_->getOriginX() != costmap_msg_->info.origin.position.x ||
+        costmap_->getOriginY() != costmap_msg_->info.origin.position.y)
+      {
+        // Update the size of the costmap
+        costmap_->resizeMap(costmap_msg_->info.width, costmap_msg_->info.height,
+          costmap_msg_->info.resolution,
+          costmap_msg_->info.origin.position.x,
+          costmap_msg_->info.origin.position.y);
+      }
 
-  unsigned char * master_array = costmap_->getCharMap();
-  unsigned int index = 0;
-  for (unsigned int i = 0; i < costmap_msg_->metadata.size_x; ++i) {
-    for (unsigned int j = 0; j < costmap_msg_->metadata.size_y; ++j) {
-      master_array[index] = costmap_msg_->data[index];
-      ++index;
+      unsigned char * master_array = costmap_->getCharMap();
+      unsigned int index = 0;
+      for (unsigned int i = 0; i < costmap_msg_->info.width; ++i) {
+        for (unsigned int j = 0; j < costmap_msg_->info.height; ++j) {
+          unsigned char value = costmap_msg_->data[index];
+          master_array[index] = interpretValue(value);
+          ++index;
+        }
+      }
     }
   }
 }
 
-void CostmapSubscriber::costmapCallback(const nav2_msgs::msg::Costmap::SharedPtr msg)
+unsigned char
+CostmapSubscriber::interpretValue(unsigned char value)
+{
+  // check if the static value is above the unknown or lethal thresholds
+  if (track_unknown_space_ && value == unknown_cost_value_) {
+    return NO_INFORMATION;
+  } else if (!track_unknown_space_ && value == unknown_cost_value_) {
+    return FREE_SPACE;
+  } else if (value >= lethal_threshold_) {
+    return LETHAL_OBSTACLE;
+  } else if (trinary_costmap_) {
+    return FREE_SPACE;
+  }
+
+  double scale = static_cast<double>(value) / lethal_threshold_;
+  return scale * LETHAL_OBSTACLE;
+}
+
+void CostmapSubscriber::costmapRawCallback(const nav2_msgs::msg::Costmap::SharedPtr msg)
+{
+  costmap_raw_msg_ = msg;
+  if (!costmap_received_) {
+    costmap_received_ = true;
+  }
+}
+
+void CostmapSubscriber::costmapCallback(const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
 {
   costmap_msg_ = msg;
   if (!costmap_received_) {
     costmap_received_ = true;
   }
 }
 
+void CostmapSubscriber::costmapUpdateCallback(
```

Comment:
This is a bug, in the situation where more than one update comes through between calls to get the costmap from the subscriber object, you leak any but the last. This callback needs to change the `costmap_` so that all updates are reflected when `getCostmap` is called

Reply:
You are right, thanks for catching. I recall when I first introduced the costmap subscriber code, we discussed that it was better to update the costmap object "on call" as opposed to every callback.

In this case, the OccupancyGridUpdate model is different, where we now need to update the costmap every callback for the updates topic. But since the first OccupancyGrid topic needs to have already created/updated the costmap before the updates come in, it might require that we update the costmap every callback as opposed to only "on call" (e.g. `getCostmap`), right?

Reply:
> require that we update the costmap every callback 

That is correct, unless you want to buffer them and do them on `getCostmap` but I think in the Cb is more appropriate

Reply:
I'm now calling to`Costmap2D()` in the callbacks directly